vm/actor/src/builtin/market/mod.rs

```diff
@@ -4,7 +4,710 @@
+                    .get(id)
+                    .map_err(|e| ActorError::new(ExitCode::ErrIllegalState, e.into()))?
+                    .ok_or_else(|| {
+                        ActorError::new(
+                            ExitCode::ErrIllegalState,
+                            "Failed to retrieve DealProposal".to_owned(),
+                        )
+                    })?;
+                assert_eq!(deal.provider, miner_addr);
+
+                let mut state: DealState = states
+                    .get(id)
+                    .map_err(|e| ActorError::new(ExitCode::ErrIllegalState, e.into()))?
+                    .ok_or_else(|| {
+                        ActorError::new(
+                            ExitCode::ErrIllegalState,
+                            "Failed to retrieve DealState".to_owned(),
+                        )
+                    })?;
+
+                // Note: we do not perform the balance transfers here, but rather simply record the flag
+                // to indicate that processDealSlashed should be called when the deferred state computation
+                // is performed. // TODO: Do that here
+
+                state.slash_epoch = OptionalEpoch(Some(rt.curr_epoch()));
+                states.set(id, state).map_err(|e| {
+                    ActorError::new(ExitCode::ErrIllegalState, format!("Set deal error: {}", e))
+                })?;
+            }
+
+            st.states = states
+                .flush()
+                .map_err(|e| ActorError::new(ExitCode::ErrIllegalState, e.into()))?;
+            Ok(())
+        })??;
+        Ok(())
+    }
+
+    fn handle_expired_deals<BS, RT>(
+        rt: &mut RT,
+        params: HandleExpiredDealsParams,
+    ) -> Result<(), ActorError>
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_type(CALLER_TYPES_SIGNABLE.iter())?;
+
+        let slashed = rt.transaction(|st: &mut State, rt| {
+            st.update_pending_deal_states(rt.store(), params.deal_ids, rt.curr_epoch())
+        })??;
+
+        // TODO: award some small portion of slashed to caller as incentive
+
+        rt.send(
+            &*BURNT_FUNDS_ACTOR_ADDR,
+            METHOD_SEND,
+            &Serialized::default(),
+            &slashed,
+        )?;
+        Ok(())
+    }
+}
+////////////////////////////////////////////////////////////////////////////////
+// Checks
+////////////////////////////////////////////////////////////////////////////////
+fn validate_deal_can_activate(
+    curr_epoch: ChainEpoch,
+    miner_addr: &Address,
+    sector_exp: ChainEpoch,
+    deal: &DealState,
+    proposal: &DealProposal,
+) -> Result<(), ActorError> {
+    if &proposal.provider != miner_addr {
+        return Err(ActorError::new(
+            ExitCode::ErrIllegalArgument,
+            "Deal has incorrect miner as its provider.".to_owned(),
+        ));
+    };
+
+    if deal.sector_start_epoch != OptionalEpoch(None) {
```

Comment:
```suggestion
    if deal.sector_start_epoch.is_some() {
```
Can use optional epoch type as `Option<ChainEpoch>` because I impl deref on it